lib/src/clojure_lsp/feature/file_management.clj

```diff
@@ -249,42 +247,42 @@
                  "Watched files analyzed, took %s"
                  (lsp.kondo/run-kondo-on-paths! filenames db* {:external? false} nil))]
     (swap! db* lsp.kondo/db-with-results result)
-    (f.diagnostic/publish-all-diagnostics! filenames components)
+    (f.diagnostic/publish-all-diagnostics! uris components)
     (producer/refresh-test-tree producer uris)
     (doseq [uri uris]
       (when (get-in @db* [:documents uri :v])
         (when-let [text (shared/slurp-uri uri)]
           (swap! db* assoc-in [:documents uri :text] text))))))
 
-(defn ^:private db-without-file [state-db uri filename]
+(defn ^:private db-without-uri [state-db uri]
   (-> state-db
-      (dep-graph/remove-file uri filename)
+      (dep-graph/remove-doc uri)
       (shared/dissoc-in [:documents uri])
-      (shared/dissoc-in [:analysis filename])
-      (shared/dissoc-in [:findings filename])))
+      (shared/dissoc-in [:analysis uri])
+      (shared/dissoc-in [:findings uri])))
 
-(defn ^:private file-deleted [{:keys [db*], :as components} uri filename]
-  (swap! db* db-without-file uri filename)
-  (f.diagnostic/publish-empty-diagnostics! uri components))
+(defn ^:private files-deleted [{:keys [db*], :as components} uris]
+  (swap! db* #(reduce db-without-uri % uris))
+  (f.diagnostic/publish-empty-diagnostics! uris components))
 
 (defn did-change-watched-files
   [changes
    {:keys [db* watched-files-chan] :as components}]
-  (doseq [{:keys [uri type]} changes]
-    (case type
-      :created (async/>!! watched-files-chan uri)
-      :changed (when (settings/get @db* [:compute-external-file-changes] true)
-                 (async/>!! watched-files-chan uri))
-      :deleted (shared/logging-task
-                 :delete-watched-file
-                 (file-deleted components uri (shared/uri->filename uri))))))
+  (let [{:keys [created changed deleted]} (group-by :type changes)
+        observe-changed? (settings/get @db* [:compute-external-file-changes] true)]
+    (doseq [created-or-changed (concat created (when observe-changed? changed))]
+      (async/>!! watched-files-chan (:uri created-or-changed)))
+    (when (seq deleted)
+      (shared/logging-task
+        :delete-watched-files
+        (files-deleted components (map :uri deleted))))))
 
 (defn did-close [uri {:keys [db*], :as components}]
   (let [filename (shared/uri->filename uri)
         source-paths (settings/get @db* [:source-paths])]
     (when (and (not (shared/external-filename? filename source-paths))
                (not (shared/file-exists? (io/file filename))))
```

Comment:
We still use filename here to decide whether a file exists on the local filesystem.

Reply:
maybe we should have a `uri-exists?` as we will end up using more and more uris and there are other places that need to do similar checks if file exists too

Reply:
Yep, I'd love to add this someday. There's another review of the code to be done to clean up this kind of thing. The dep-graph includes `:internal?` so we might be able to reduce the number of uses of `external-filename?` too.